src/main/java/com/google/cloud/spanner/pgadapter/utils/MutationWriter.java

```diff
@@ -14,225 +14,580 @@
+        connection,
+        null,
+        tableName,
+        tableColumns,
+        indexedColumnsCount,
+        format,
+        hasHeader);
+  }
+
+  @VisibleForTesting
+  MutationWriter(
+      CopyTransactionMode transactionMode,
+      CloudSpannerJdbcConnection connection,
+      DatabaseClient databaseClient,
+      String tableName,
+      Map<String, TypeCode> tableColumns,
+      int indexedColumnsCount,
+      CSVFormat format,
+      boolean hasHeader)
+      throws IOException {
+    this.transactionMode = transactionMode;
+    this.connection = connection;
+    this.databaseClient = databaseClient;
     this.hasHeader = hasHeader;
     this.isHeaderParsed = false;
     this.tableName = tableName;
     this.tableColumns = tableColumns;
+    int mutationLimit =
+        Math.max(
+            Integer.parseInt(
+                System.getProperty(
+                    "copy_in_mutation_limit", String.valueOf(DEFAULT_MUTATION_LIMIT))),
+            1);
+    this.maxBatchSize = Math.max(mutationLimit / (tableColumns.size() + indexedColumnsCount), 1);
+    float commitLimitMultiplierFactor =
+        Math.max(
+            Float.parseFloat(
+                System.getProperty(
+                    "copy_in_commit_limit_multiplier_factor",
+                    String.valueOf(DEFAULT_COMMIT_LIMIT_MULTIPLIER_FACTOR))),
+            1.0f);
+    this.commitSizeLimitForBatching =
+        Math.round((float) commitSizeLimit / commitLimitMultiplierFactor);
+    this.maxParallelism =
+        Math.max(
+            Integer.parseInt(
+                System.getProperty(
+                    "copy_in_max_parallelism", String.valueOf(DEFAULT_MAX_PARALLELISM))),
+            1);
+    this.pipeBufferSize =
+        Math.max(
+            Integer.parseInt(
+                System.getProperty(
+                    "copy_in_pipe_buffer_size", String.valueOf(DEFAULT_PIPE_BUFFER_SIZE))),
+            1024);
     this.format = format;
-    this.mutations = new ArrayList<>();
-  }
-
-  /** @return list of CopyData mutations not yet written into Spanner */
-  public List<Mutation> getMutations() {
-    return this.mutations;
+    this.parser = createParser();
   }
 
   /** @return number of rows copied into Spanner */
-  public int getRowCount() {
+  public long getRowCount() {
     return this.rowCount;
   }
 
-  public void addCopyData(ConnectionHandler connectionHandler, byte[] payload) throws Exception {
-    this.payload.write(payload, 0, payload.length);
-    if (!commitSizeIsWithinLimit()) {
-      handleError(connectionHandler);
-      throw new SQLException(
-          "Commit size: " + this.payload.size() + " has exceeded the limit: " + COMMIT_LIMIT);
+  public void addCopyData(byte[] payload) throws SQLException {
+    synchronized (lock) {
+      if (this.exception != null) {
+        throw this.exception;
+      }
+    }
+    try {
+      this.payload.write(payload);
+    } catch (IOException e) {
+      throw new SQLException(e);
```

Comment:
This just keeps the number of different types of exceptions that we need to handle further down the chain smaller. (And it will be replaced with `SpannerException` once we have switched to the Connection API)